File: app/src/lib/wire.ts
/**
 * The JSON `clob-stream` actually sends, spelled exactly as it arrives.
 *
 * Separate from the decoded types in {@link ./decode.ts} on purpose. These say `string`
 * where the indexer sends a quoted number, because that is what `JSON.parse` produces —
 * writing `bigint` here would describe a value nobody ever holds and move the conversion
 * somewhere it could be forgotten.
 *
 * The indexer's rule, from its README: money and identities are strings, coordinates and
 * tallies are numbers. A `u64` price cannot be a JSON number — a bid's stored sequence
 * number sits just below 2^64, and a double loses it. Slots and seat indices can, being
 * bounded far below 2^53.
 */

/** A price level as sent. */
export interface WireLevel {
  price_in_ticks: string;
  base_lots: string;
}

/** Tick and lot geometry, carried on every market summary. */
export interface WireLots {
  base_lots_per_base_unit: string;
  tick_size_in_quote_lots_per_base_unit: string;
  base_atoms_per_base_lot: string;
  quote_atoms_per_quote_lot: string;
}

/** One market, as `/v1/markets` returns it. */
export interface WireMarketSummary {
  market: string;
  slot: number;
  finalized_through: number;
  base_mint: string;
  quote_mint: string;
  base_vault: string;
  quote_vault: string;
  vault_signer: string;
  taker_fee_bps: number;
  lots: WireLots;
  best_bid_in_ticks: string | null;
  best_ask_in_ticks: string | null;
  spread_in_ticks: string | null;
  mid_price_in_ticks: string | null;
  last_price_in_ticks: string | null;
  bid_orders: number;
  ask_orders: number;
  base_lots_deposited: string;
  quote_lots_deposited: string;
  seats: number;
  trades_seen: number;
}

/** One market's book, as `/book` returns it. */
export interface WireBook {
  market: string;
  slot: number;
  bids: WireLevel[];
  asks: WireLevel[];
  taker_fee_bps: number;
  finalized_through: number;
}

/** A live fill. */
export interface WireTrade {
  slot: number;
  price_in_ticks: string;
  base_lots: string;
  quote_lots: string;
  taker_side: "bid" | "ask";
  maker_seat: number;
  taker_seat: number | null;
  finalized: boolean;
}

/** A stored fill, as `/history` returns it. Always rooted, so it carries no flag. */
export interface WireHistoricalTrade {
  slot: number;
  signature: string;
  price_in_ticks: string;
  base_lots: string;
  quote_lots: string;
  taker_side: "bid" | "ask";
  maker_seat: number;
  taker_seat: number | null;
}

/** Trading activity over a span of slots, as `/window` returns it. */
export interface WireWindow {
  market: string;
  from_slot: number;
  to_slot: number;
  slots: number;
  open_in_ticks: string | null;
  high_in_ticks: string | null;
  low_in_ticks: string | null;
  close_in_ticks: string | null;
  change_in_ticks: string | null;
  vwap_in_ticks: string | null;
  base_lots: string;
  quote_lots: string;
  trades: number;
  truncated: boolean;
}

/** One of a trader's resting orders. */
export interface WireOpenOrder {
  side: "bid" | "ask";
  price_in_ticks: string;
  /** What `CancelOrder` takes. Side-encoded — not the arrival order. */
  order_sequence_number: string;
  /** Arrival order, decoded. What the tape records, so the field to join a fill against. */
  sequence_number: string;
  base_lots: string;
}

/** A trader's position in one market, as `/traders/{wallet}` returns it. */
export interface WireTraderView {
  market: string;
  trader: string;
  seat: number;
  slot: number;
  finalized_through: number;
  base_lots_free: string;
  base_lots_locked: string;
  quote_lots_free: string;
  quote_lots_locked: string;
  orders: WireOpenOrder[];
}

/** One OHLCV bucket, keyed by slot rather than by wall clock. */
export interface WireCandle {
  start_slot: number;
  open: string;
  high: string;
  low: string;
  close: string;
  base_lots: string;
  quote_lots: string;
  trades: number;
}

/**
 * A message from the live feed.
 *
 * Four kinds, and the two that are easy to skip are the two that matter. `retract` says
 * trades already shown did not happen, because the slot that produced them was abandoned.
 * `lagged` says this subscriber fell behind and messages were dropped — silence and a gap
 * look identical, so the feed says which it is.
 */
export type WireMessage =
  | {
      type: "snapshot";
      market: string;
      slot: number;
      finalized_through: number;
      bids: WireLevel[];
      asks: WireLevel[];
    }
  | {
      type: "update";
      slot: number;
      trades: WireTrade[];
      bids: WireLevel[];
      asks: WireLevel[];
      best_bid: string | null;
      best_ask: string | null;
      finalized_through: number;
    }
  | { type: "retract"; slot: number; trades: number }
  | { type: "lagged"; missed: number };

File: crates/clob-tui/src/wire.rs
//! The JSON `clob-stream` sends, spelled exactly as it arrives.
//!
//! # Why every quantity is a string here
//!
//! The indexer quotes money and order identities as decimal strings because JSON has one
//! numeric type — an IEEE-754 double — and a `u64` does not fit in one. A bid's stored
//! sequence number is the complement of the arrival counter, so it sits just below
//! `u64::MAX`, and a JavaScript client reading it as a number would cancel an order that
//! does not exist.
//!
//! That constraint is not ours. Rust parses these back to `u64` exactly, so the round trip
//! is lossless in both directions and the strings cost one `parse` per field. They are kept
//! as `String` in these types rather than deserialised straight to `u64` so that a
//! malformed field fails where it is read, with the field named, rather than failing the
//! whole frame with serde's message about the wrong type.

use serde::Deserialize;

/// One aggregated price level.
#[derive(Clone, Debug, Deserialize)]
pub struct Level {
    pub price_in_ticks: String,
    pub base_lots: String,
}

/// Tick and lot geometry, carried on every market summary.
#[derive(Clone, Debug, Deserialize)]
pub struct Lots {
    pub base_lots_per_base_unit: String,
    pub tick_size_in_quote_lots_per_base_unit: String,
    pub base_atoms_per_base_lot: String,
    pub quote_atoms_per_quote_lot: String,
}

/// One market, as `/v1/markets` returns it.
#[derive(Clone, Debug, Deserialize)]
pub struct MarketSummary {
    pub market: String,
    pub slot: u64,
    pub finalized_through: u64,
    pub base_mint: String,
    pub quote_mint: String,
    pub taker_fee_bps: u64,
    pub lots: Lots,
    pub best_bid_in_ticks: Option<String>,
    pub best_ask_in_ticks: Option<String>,
    pub spread_in_ticks: Option<String>,
    pub mid_price_in_ticks: Option<String>,
    pub last_price_in_ticks: Option<String>,
    pub bid_orders: usize,
    pub ask_orders: usize,
    pub seats: usize,
    pub trades_seen: u64,
}

/// One fill.
#[derive(Clone, Debug, Deserialize)]
pub struct Trade {
    pub slot: u64,
    pub price_in_ticks: String,
    pub base_lots: String,
    pub taker_side: String,
    pub maker_seat: u32,
    pub taker_seat: Option<u32>,
}

/// One of a trader's resting orders.
#[derive(Clone, Debug, Deserialize)]
pub struct OpenOrder {
    pub side: String,
    pub price_in_ticks: String,
    pub base_lots: String,
}

/// A trader's position in one market.
#[derive(Clone, Debug, Deserialize)]
pub struct TraderView {
    pub seat: u32,
    pub base_lots_free: String,
    pub base_lots_locked: String,
    pub quote_lots_free: String,
    pub quote_lots_locked: String,
    pub orders: Vec<OpenOrder>,
}

/// A message from the live feed.
///
/// Four kinds, and the two that are easy to skip are the two that matter. `retract` says
/// fills already on screen did not happen, because the slot that produced them was
/// abandoned. `lagged` says this subscriber fell behind and messages were dropped — silence
/// and a gap look identical, so the feed says which it is.
#[derive(Clone, Debug, Deserialize)]
#[serde(tag = "type", rename_all = "snake_case")]
pub enum Message {
    Snapshot {
        slot: u64,
        finalized_through: u64,
        bids: Vec<Level>,
        asks: Vec<Level>,
    },
    Update {
        slot: u64,
        trades: Vec<Trade>,
        bids: Vec<Level>,
        asks: Vec<Level>,
        finalized_through: u64,
    },
    Retract {
        slot: u64,
        trades: usize,
    },
    Lagged {
        missed: u64,
    },
}
